include/glow/ExecutionEngine/ExecutionEngine2.h

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_EXECUTIONENGINE_EXECUTIONENGINE2_H
+#define GLOW_EXECUTIONENGINE_EXECUTIONENGINE2_H
+
+#include "glow/Backend/Backend.h"
+#include "glow/Backend/CompiledFunction.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Base/Train.h"
+#include "glow/Base/Traits.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/PlaceholderBindings.h"
+#include "glow/Runtime/HostManager/HostManager.h"
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include <memory>
+#include <unordered_map>
+
+namespace glow {
+
+/// This is the ExecutionEngine. It owns the Graph, the backend, and the
+/// compiled function.  The Graph, etc in this class are defined as pointers, in
+/// order to erase the type and prevent the internal types from leaking out to
+/// the users of this class.
+class ExecutionEngine final {
+  /// Module containing the function and supporting information.
+  std::unique_ptr<Module> module_;
+
+  /// Raw pointer to module_ this is to support module access after the module
+  /// has been added to hostManager_.
+  Module *rawModule_;
```

Comment:
Ick :( I know this is due to the existing interfaces and you don't have alternatives but this is a sign we don't want to use unique_ptr for module.